src/main/resources/hudson/plugins/sidebar_link/SidebarLinkPlugin/upload.js

```diff
@@ -0,0 +1,25 @@
+Behaviour.specify("#sidebar-link-upload-button", "sidebar-link-upload", 0, function(e) {
```

Comment:
I'm not happy to merge this JS code, prefer to have ready to use library

Reply:
I believe this is not the only plugin that uploads the file and I believe there is one common reusable approach for Jenkins like for other controls

Reply:
Usually files are uploaded during a form submit (e.g. when you have a build with a file parameter and you start a build).
But obviously this is not desired here as you want to upload the file and then be able to configure it without the need to press apply or save. Otherwise the existing approach with the iframe would not be here.
This iframe is not really the way to go in my eyes. After uploading you get a simple html snippet that contains no styling and you need to press back to get back to the place where you can upload files.
Of course you can include the file upload in every submit of the Jenkins config though that seems also not the right approach for me.

Reply:
@daniel-beck do we have reusable component for file upload instead of having this custom JavaScript code ?

Reply:
Not that I'm aware of.

Reply:
how about some opensource library eg https://github.com/pqina/filepond

Reply:
Jenkins is using pure <input> field for uploading the pluign

Reply:
Feel free to use whatever you want in your plugin.

For something provided by core, to be generally usable by plugins, it should probably come in the form of a Jelly taglib to abstract away the implementation and future-proof it. So that'll require a bit of work, and this hasn't really come up before AFAIK.

Reply:
I think using such a library is overkill. And it will not prevent that you need some javascript to connect the file input field with the library.  Also for the upload you need to include a crumb, so in the end I doubt that when using the library you will get less javascript code.
And half of the code that we have now is related to notifying the user of the result.
Having a taglib might be useful, but I'm not aware of any other use case like the one here, where the upload should not be done as part of a form submission using the save or apply buttons.
If you want I can create a jelly taglib here in the plugin.

Reply:
> If you want I can create a jelly taglib here in the plugin.

Sounds like good plan @mawinter69 

I also found https://github.com/jenkinsci/jenkins/pull/7452